File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.cpp
class Solution {
public:
    void bfs(int node, vector<int>& dist, vector<int>& edges) {
        unordered_set<int> vis;

        queue<int> q;
        q.push(node);
        vis.insert(node);
        int lvl = 0;
        while (!q.empty()) {
            int sz = q.size();

            for (int i = 0; i < sz; i++) {
                int curr = q.front();
                q.pop();
                if(dist[curr]==-1) dist[curr]=lvl;

                int nextNode=edges[curr];
                if(nextNode!=-1 && vis.find(nextNode)==vis.end()){
                    q.push(nextNode);
                    vis.insert(nextNode);
                }
            }

            lvl++;
        }

        return;
    }
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        int n = edges.size();

        vector<int> dist1(n + 1, -1), dist2(n + 1, -1);
        bfs(node1,dist1,edges);
        bfs(node2,dist2,edges);

        // for(auto it:dist1) cout<<it<<" ";
        // cout<<endl;
        // for(auto it:dist2) cout<<it<<" ";
        // cout<<endl;
        int minDist=INT_MAX; int node=-1;
        for(int i=0;i<=n;i++){
            if(dist1[i]!=-1 && dist2[i]!=-1){
                if(minDist>max(dist1[i],dist2[i])){
                    node=i;
                    minDist=max(dist1[i],dist2[i]);
                }
            }
        }
        return node;
        // start dfs from node1
    }
};

File: 2359-find-closest-node-to-given-two-nodes/2359-find-closest-node-to-given-two-nodes.java
class Solution {
    public int closestMeetingNode(int[] edges, int node1, int node2) {

        int n = edges.length;
        int[] n1_dist = new int[n], n2_dist = new int[n];
        for (int i = 0; i < n; i++) {
            n1_dist[i] = Integer.MAX_VALUE;
            n2_dist[i] = Integer.MAX_VALUE;
        }
        n1_dist[node1]=0; n2_dist[node2]=0;
        Queue<Integer> q= new LinkedList<>();

        //for node 1
        q.offer(node1);
        while(!q.isEmpty()){
            int node=q.poll();
            int distance=n1_dist[node];
            int destNode=edges[node];
            
            if(destNode==-1) continue; //end of graph
            else{
                if(n1_dist[destNode]<=distance+1) continue;
                else{
                    n1_dist[destNode]=distance+1;
                    q.offer(destNode);
                }
            }

        }

        //for second node
        q.offer(node2);
        while(!q.isEmpty()){
            int node=q.poll();
            int distance=n2_dist[node];
            int destNode=edges[node];
            
            if(destNode==-1) continue; //end of graph
            else{
                if(n2_dist[destNode]<=distance+1) continue;
                else{
                    n2_dist[destNode]=distance+1;
                    q.offer(destNode);
                }
            }
        }


        //find min distance
        int minDist=Integer.MAX_VALUE;
        int minIndex=-1;
        for(int i=0;i<n;i++){
            // System.out.println(n1_dist[i]+" "+n2_dist[i]);
            if(n1_dist[i]!=Integer.MAX_VALUE && n2_dist[i]!=Integer.MAX_VALUE){

                if(minDist>Math.min(minDist,Math.max(n1_dist[i],n2_dist[i]))){
                    minDist=Math.min(minDist,Math.max(n1_dist[i],n2_dist[i]));
                    minIndex=i;
                }
            }
        }

        return minIndex;
    }
}